// starling-developer-sdk/dist/components/entities/payee.js
(function (global, factory) {
  if (typeof define === "function" && define.amd) {
    define(["axios", "debug", "../utils/http", "../utils/validator"], factory);
  } else if (typeof exports !== "undefined") {
    factory(require("axios"), require("debug"), require("../utils/http"), require("../utils/validator"));
  } else {
    var mod = {
      exports: {}
    };
    factory(global.axios, global.debug, global.http, global.validator);
    global.payee = mod.exports;
  }
})(typeof globalThis !== "undefined" ? globalThis : typeof self !== "undefined" ? self : this, function (_axios, _debug, _http, _validator) {
  "use strict";

  _axios = _interopRequireDefault(_axios);
  _debug = _interopRequireDefault(_debug);

  function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

  function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

  function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

  function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

  function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

  function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

  function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

  var log = (0, _debug["default"])('starling:payee-service');
  /**
   * Service to interact with an account holder's payees
   */

  var Payee = /*#__PURE__*/function () {
    /**
     * Create a new payee service
     * @param {Object} options - configuration parameters
     */
    function Payee(options) {
      _classCallCheck(this, Payee);

      this.options = options;
    }
    /**
     * Get an account holder's payees
     * @param {string} parameters.apiUrl - the API URL
     * @param {string} parameters.accessToken - the oauth bearer token.
     * @return {Promise} - the http request promise
     */


    _createClass(Payee, [{
      key: "getPayees",
      value: function getPayees(parameters) {
        parameters = Object.assign({}, this.options, parameters);
        (0, _validator.minAPIParameterValidator)(parameters);
        var _parameters = parameters,
            apiUrl = _parameters.apiUrl,
            accessToken = _parameters.accessToken;
        var url = "".concat(apiUrl, "/api/v2/payees");
        log("GET ".concat(url));
        return (0, _axios["default"])({
          method: 'GET',
          url: url,
          headers: (0, _http.defaultHeaders)(accessToken)
        });
      }
      /**
       * Create a new payee
       * @param {string} parameters.apiUrl - the API URL
       * @param {string} parameters.accessToken - the oauth bearer token.
       * @param {Object} parameters.payeeCreationRequest - the payee creation request.
       * @return {Promise} - the http request promise
       */

    }, {
      key: "createPayee",
      value: function createPayee(parameters) {
        parameters = Object.assign({}, this.options, parameters);
        createPayeeParameterValidator(parameters);
        var _parameters2 = parameters,
            apiUrl = _parameters2.apiUrl,
            accessToken = _parameters2.accessToken,
            payeeCreationRequest = _parameters2.payeeCreationRequest;
        var url = "".concat(apiUrl, "/api/v2/payees");
        log("PUT ".concat(url));
        return (0, _axios["default"])({
          method: 'PUT',
          url: url,
          headers: (0, _http.payloadHeaders)(accessToken),
          data: JSON.stringify(payeeCreationRequest)
        });
      }
      /**
       * Delete an account holder's payee
       * @param {string} parameters.apiUrl - the API URL
       * @param {string} parameters.accessToken - the oauth bearer token.
       * @param {string} parameters.payeeUid - the payeeUid of the payee to be deleted.
       * @return {Promise} - the http request promise
       */

    }, {
      key: "deletePayee",
      value: function deletePayee(parameters) {
        parameters = Object.assign({}, this.options, parameters);
        deletePayeeParameterValidator(parameters);
        var _parameters3 = parameters,
            apiUrl = _parameters3.apiUrl,
            accessToken = _parameters3.accessToken,
            payeeUid = _parameters3.payeeUid;
        var url = "".concat(apiUrl, "/api/v2/payees/").concat(payeeUid);
        log("DELETE ".concat(url));
        return (0, _axios["default"])({
          method: 'DELETE',
          url: url,
          headers: (0, _http.defaultHeaders)(accessToken)
        });
      }
    }]);

    return Payee;
  }();

  var createPayeeParameterValidator = _validator.struct["interface"](_objectSpread(_objectSpread({}, _validator.minAPIParameterDefintion), {}, {
    payeeCreationRequest: _validator.struct.object({
      payeeName: 'string',
      phoneNumber: 'string?',
      payeeType: _validator.struct["enum"](['INDIVIDUAL', 'BUSINESS']),
      firstName: 'string?',
      middleName: 'string?',
      lastName: 'string?',
      businessName: 'string?',
      dateOfBirth: 'date?',
      accounts: _validator.struct.optional([_validator.struct.object({
        description: 'string',
        defaultAccount: 'boolean',
        countryCode: 'string',
        accountIdentifier: 'string',
        bankIdentifier: 'string',
        bankIdentifierType: _validator.struct["enum"](['SORT_CODE', 'SWIFT', 'IBAN', 'ABA', 'ABA_WIRE', 'ABA_ACH'])
      })])
    })
  }));

  var deletePayeeParameterValidator = _validator.struct["interface"](_objectSpread(_objectSpread({}, _validator.minAPIParameterDefintion), {}, {
    payeeUid: 'uuid'
  }));

  module.exports = Payee;
});
