// starling-developer-sdk/dist/components/entities/payment.js
(function (global, factory) {
  if (typeof define === "function" && define.amd) {
    define(["axios", "debug", "../utils/http", "../utils/validator"], factory);
  } else if (typeof exports !== "undefined") {
    factory(require("axios"), require("debug"), require("../utils/http"), require("../utils/validator"));
  } else {
    var mod = {
      exports: {}
    };
    factory(global.axios, global.debug, global.http, global.validator);
    global.payment = mod.exports;
  }
})(typeof globalThis !== "undefined" ? globalThis : typeof self !== "undefined" ? self : this, function (_axios, _debug, _http, _validator) {
  "use strict";

  _axios = _interopRequireDefault(_axios);
  _debug = _interopRequireDefault(_debug);

  function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

  function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

  function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

  function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

  function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

  function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

  function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

  var log = (0, _debug["default"])('starling:payment-service');
  /**
   * Service to interact with a customer's payments
   */

  var Payment = /*#__PURE__*/function () {
    /**
     * Create a new payment service
     * @param {Object} options - configuration parameters
     */
    function Payment(options) {
      _classCallCheck(this, Payment);

      this.options = options;
    }
    /**
     * Get a payment order
     * @param {string} parameters.apiUrl - the API URL
     * @param {string} parameters.accessToken - the oauth bearer token
     * @param {string} parameters.paymentOrderUid - the payment order uid
     * @return {Promise} - the http request promise
     */


    _createClass(Payment, [{
      key: "getPaymentOrder",
      value: function getPaymentOrder(parameters) {
        parameters = Object.assign({}, this.options, parameters);
        getPaymentOrderParameterValidator(parameters);
        var _parameters = parameters,
            apiUrl = _parameters.apiUrl,
            accessToken = _parameters.accessToken,
            paymentOrderUid = _parameters.paymentOrderUid;
        var url = "".concat(apiUrl, "/api/v2/payments/local/payment-order/").concat(paymentOrderUid);
        log("GET ".concat(url));
        return (0, _axios["default"])({
          method: 'GET',
          url: url,
          headers: (0, _http.defaultHeaders)(accessToken)
        });
      }
      /**
       * Get a payment order's payments
       * @param {string} parameters.apiUrl - the API URL
       * @param {string} parameters.accessToken - the oauth bearer token
       * @param {string} parameters.paymentOrderUid - the payment order uid
       * @return {Promise} - the http request promise
       */

    }, {
      key: "getPaymentOrderPayments",
      value: function getPaymentOrderPayments(parameters) {
        parameters = Object.assign({}, this.options, parameters);
        getPaymentOrderPaymentsParameterValidator(parameters);
        var _parameters2 = parameters,
            apiUrl = _parameters2.apiUrl,
            accessToken = _parameters2.accessToken,
            paymentOrderUid = _parameters2.paymentOrderUid;
        var url = "".concat(apiUrl, "/api/v2/payments/local/payment-order/").concat(paymentOrderUid, "/payments");
        log("GET ".concat(url));
        return (0, _axios["default"])({
          method: 'GET',
          url: url,
          headers: (0, _http.defaultHeaders)(accessToken)
        });
      }
      /**
       * List standing orders
       * @param {string} parameters.apiUrl - the API URL
       * @param {string} parameters.accessToken - the oauth bearer token
       * @param {string} parameters.accountUid - the account uid of the account to get standing orders of
       * @param {string} parameters.categoryUid - the category uid of the category to get standing orders of
       * @return {Promise} - the http request promise
       */

    }, {
      key: "listStandingOrders",
      value: function listStandingOrders(parameters) {
        parameters = Object.assign({}, this.options, parameters);
        listStandingOrdersParameterValidator(parameters);
        var _parameters3 = parameters,
            apiUrl = _parameters3.apiUrl,
            accessToken = _parameters3.accessToken,
            accountUid = _parameters3.accountUid,
            categoryUid = _parameters3.categoryUid;
        var url = "".concat(apiUrl, "/api/v2/payments/local/account/").concat(accountUid, "/category/").concat(categoryUid, "/standing-orders");
        log("GET ".concat(url));
        return (0, _axios["default"])({
          method: 'GET',
          url: url,
          headers: (0, _http.defaultHeaders)(accessToken)
        });
      }
      /**
       * Get a standing order
       * @param {string} parameters.apiUrl - the API URL
       * @param {string} parameters.accessToken - the oauth bearer token
       * @param {string} parameters.accountUid - the account uid of the standing order
       * @param {string} parameters.categoryUid - the category uid of the standing order
       * @param {string} parameters.paymentOrderUid - the payment order uid of the standing order
       * @return {Promise} - the http request promise
       */

    }, {
      key: "getStandingOrder",
      value: function getStandingOrder(parameters) {
        parameters = Object.assign({}, this.options, parameters);
        getStandingOrderParameterValidator(parameters);
        var _parameters4 = parameters,
            apiUrl = _parameters4.apiUrl,
            accessToken = _parameters4.accessToken,
            accountUid = _parameters4.accountUid,
            categoryUid = _parameters4.categoryUid,
            paymentOrderUid = _parameters4.paymentOrderUid;
        var url = "".concat(apiUrl, "/api/v2/payments/local/account/").concat(accountUid, "/category/").concat(categoryUid, "/standing-orders/").concat(paymentOrderUid);
        log("GET ".concat(url));
        return (0, _axios["default"])({
          method: 'GET',
          url: url,
          headers: (0, _http.defaultHeaders)(accessToken)
        });
      }
    }]);

    return Payment;
  }();

  var getPaymentOrderParameterValidator = _validator.struct["interface"](_objectSpread(_objectSpread({}, _validator.minAPIParameterDefintion), {}, {
    paymentOrderUid: 'uuid'
  }));

  var getPaymentOrderPaymentsParameterValidator = _validator.struct["interface"](_objectSpread(_objectSpread({}, _validator.minAPIParameterDefintion), {}, {
    paymentOrderUid: 'uuid'
  }));

  var listStandingOrdersParameterValidator = _validator.struct["interface"](_objectSpread(_objectSpread({}, _validator.minAPIParameterDefintion), {}, {
    accountUid: 'uuid',
    categoryUid: 'uuid'
  }));

  var getStandingOrderParameterValidator = _validator.struct["interface"](_objectSpread(_objectSpread({}, _validator.minAPIParameterDefintion), {}, {
    accountUid: 'uuid',
    categoryUid: 'uuid',
    paymentOrderUid: 'uuid'
  }));

  module.exports = Payment;
});
