// starling-developer-sdk/dist/components/starling.js
(function (global, factory) {
  if (typeof define === "function" && define.amd) {
    define(["./entities/account", "./entities/accountHolder", "./entities/address", "./entities/feedItem", "./entities/card", "./entities/oauth", "./entities/payee", "./entities/payment", "./entities/mandate", "./entities/savingsGoal", "./entities/identity"], factory);
  } else if (typeof exports !== "undefined") {
    factory(require("./entities/account"), require("./entities/accountHolder"), require("./entities/address"), require("./entities/feedItem"), require("./entities/card"), require("./entities/oauth"), require("./entities/payee"), require("./entities/payment"), require("./entities/mandate"), require("./entities/savingsGoal"), require("./entities/identity"));
  } else {
    var mod = {
      exports: {}
    };
    factory(global.account, global.accountHolder, global.address, global.feedItem, global.card, global.oauth, global.payee, global.payment, global.mandate, global.savingsGoal, global.identity);
    global.starling = mod.exports;
  }
})(typeof globalThis !== "undefined" ? globalThis : typeof self !== "undefined" ? self : this, function (_account, _accountHolder, _address, _feedItem, _card, _oauth, _payee, _payment, _mandate, _savingsGoal, _identity) {
  "use strict";

  _account = _interopRequireDefault(_account);
  _accountHolder = _interopRequireDefault(_accountHolder);
  _address = _interopRequireDefault(_address);
  _feedItem = _interopRequireDefault(_feedItem);
  _card = _interopRequireDefault(_card);
  _oauth = _interopRequireDefault(_oauth);
  _payee = _interopRequireDefault(_payee);
  _payment = _interopRequireDefault(_payment);
  _mandate = _interopRequireDefault(_mandate);
  _savingsGoal = _interopRequireDefault(_savingsGoal);
  _identity = _interopRequireDefault(_identity);

  function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

  function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

  /**
   * @property {Identity} identity
   * @property {AccountHolder} accountHolder
   * @property {Account} account
   * @property {Address} address
   * @property {FeedItem} feedItem
   * @property {Payment} payment
   * @property {Mandate} mandate
   * @property {Payee} payee
   * @property {Card} card
   * @property {SavingsGoal} savingsGoal
   * @property {OAuth} oAuth
   */
  var Starling =
  /**
   * Create an instance of the starling client
   * @param {Object=} options - configuration parameters
   */
  function Starling(options) {
    _classCallCheck(this, Starling);

    var defaults = {
      apiUrl: 'https://api.starlingbank.com',
      clientId: '',
      clientSecret: ''
    };
    this.config = Object.assign({}, defaults, options);
    this.identity = new _identity["default"](this.config);
    this.accountHolder = new _accountHolder["default"](this.config);
    this.account = new _account["default"](this.config);
    this.address = new _address["default"](this.config);
    this.feedItem = new _feedItem["default"](this.config);
    this.payment = new _payment["default"](this.config);
    this.mandate = new _mandate["default"](this.config);
    this.payee = new _payee["default"](this.config);
    this.card = new _card["default"](this.config);
    this.savingsGoal = new _savingsGoal["default"](this.config);
    this.oAuth = new _oauth["default"](this.config);
  };

  module.exports = Starling;
});
